### PR TITLE
Adding openshift.patch/selector.patch and refactoring

### LINKED ISSUE #62: natively support the patch command
Add openshift.patch and <selector>.patch support to the OpenShift DSL

### PR DESCRIPTION
**Pull request by coreydaley**
Adds openshift.patch() and <selector>.patch commands (Fixes #62)

Reformats and refactors various bits of code and documentation (including line endings and extraneous white space)

### CODE REVIEW

**Comment by gabemontero** [comment]
[test]

File: src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy [hunk 1]
```diff
@@ -514,106 +496,137 @@ class OpenShiftDSL implements Serializable {
...
         String s = obj.toString();
         boolean markup = s.contains("{") || s.contains(":"); // does this look like json or yaml?
         boolean httpref = s.toLowerCase().startsWith("http") && verb.equalsIgnoreCase("create"); // a create from a http or https raw.githubuser path
 
-        Result r = new Result( verb );
+        Result r = new Result(verb);
 
-        if ( markup && !httpref) {
-            FilePath f = currentContext.exec.getWorkspaceFilePath().createTextTempFile( verb, ".markup", s, false );
+        if (markup) {
```
Comment (gabemontero) [inline src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy]: so an http url will always have a `:` in it, so `markup` will always be true if `httpref` is true

which means we'll never hit the `else if (httpref)` block

seems like this needs more refactoring

File: src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy [hunk 2]
```diff
@@ -514,106 +496,137 @@ class OpenShiftDSL implements Serializable {
...
         }
 
         String s = obj.toString();
         boolean markup = s.contains("{") || s.contains(":")
         boolean httpref = s.toLowerCase().startsWith("http")
-        Result r = new Result( "process" )
-        if ( markup && !httpref ) { // does this look like json or yaml?
-            FilePath f = currentContext.exec.getWorkspaceFilePath().createTextTempFile( "process", ".markup", s, false );
+        Result r = new Result("process")
+        if (markup) { // does this look like json or yaml?
```
Comment (gabemontero) [inline src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy]: same situation here

File: src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy [hunk 3]
```diff
@@ -514,106 +496,137 @@ class OpenShiftDSL implements Serializable {
...
+
+        if (obj instanceof Map) {
+            obj = JsonOutput.toJson(obj);
+        }
+
+        String s = obj.toString();
+        boolean markup = s.contains("{") || s.contains(":")
+        boolean httpref = s.toLowerCase().startsWith("http")
+        Result r = new Result("patch")
+        if (markup) { // does this look like json or yaml?
```
Comment (gabemontero) [inline src/main/resources/com/openshift/jenkins/plugins/OpenShiftDSL.groovy]: same here

**Comment by coreydaley** [comment]
[test]

**Comment by coreydaley** [comment]
@gabemontero ptal

**Comment by coreydaley** [comment]
@gabemontero ptal when you get a chance

**Comment by gabemontero** [comment]
hey @coreydaley 

I'll merge this week ... just want to get a better feel first on 3.7 stop ship, when it goes out, etc. as this change should be post 3.7 ... I'd like to keep the commits in sync between the openshift and jenkinsci versions of this repo if possible.

**Comment by coreydaley** [comment]
Sounds good, thanks Gabe

**Comment by gabemontero** [comment]
[merge]

**Pull request merged**